File: ssd.py
import math;
import sys;
import ast;
import re;
import random;

"""
SSD Commands:
    write <address> <data>
        writes the data to memory at the given address
        address: needs to be an integer between 0 and MEMORY_CAPACITY
        data: needs to be in the format [1,2,3] with no spaces
    read <address> <length>
        reads and returns data of a specified length at a specified address
        address: needs to be an integer between 0 and MEMORY_CAPACITY
        length: needs to be an integer and the read can't run over MEMORY_CAPACITY
    print [all | memory | state | map]
        'print all' or 'p' prints everything
        'print memory' or 'print mem' prints just the physical blocks
        'print state' prints physical_page_state
        'print map' prints the page_map
    exit
        'exit' or 'e' or 'quit' or 'q' exits the program
    garbage
        runs the garbage collection process
    expect <address> <length> <data>
        reads data of a specified length from a specified address and compares
            it to the given data. Quits the program if the expectation fails
        address: needs to be an integer between 0 and MEMORY_CAPACITY
        length: needs to be an integer and the read can't run over MEMORY_CAPACITY
        data: needs to be in the format [1,2,3] with no spaces
"""

#TODO: more efficient garbage collection
#TODO: wear-tracking (make counts of how many times things have been erased)
#TODO: wear-leveling (the find_available_page function needs to find the right pages)
#TODO: reading memory that hasn't been written to should probably return all zeros very quickly
#TODO: page_state_store and such should maybe be ranges at a time. Not sure we even care about these.
#TODO: error handling when memory is full
#TODO: make separate storage files and allow user to specify (not urgent)
#TODO: serialize memory array and maps

# parameters
NUM_BLOCKS = 4;
PAGES_PER_BLOCK = 4;# = 128;
BYTES_PER_PAGE = 8;# = 2000;

# convenience variables
BYTES_PER_BLOCK = None;
MEMORY_CAPACITY = None;

# class to represent an enum (since enums were not untroduced until Python 3)
class Enum(set):
    def __getattr__(self, name):
        if name in self:
            return name;
        else:
            raise AttributeError;

# enum for possible states of a physical memory page
PageState = Enum(['AVAILABLE', 'IN_USE', 'INVALID']);

# the actual array that represents the physical memory
blocks = None;

# keeps track of whether each page is AVAILABLE, IN_USE, or INVALID
physical_page_state = None;

# maps virtual page locations to where they are being stored
page_map = None;

# holds a trace of the actions that an operation incurred
log = list();

# performs a read from the memory of arbitrary length starting from any byte address
def read(address, length):
    read_data = list();
    check_address('memory', address, length);
    while length > 0:
        block_address = address // BYTES_PER_BLOCK;
        byte_address_in_block = address % BYTES_PER_BLOCK;
        page_address = byte_address_in_block // BYTES_PER_PAGE;
        byte_address_in_page = byte_address_in_block % BYTES_PER_PAGE;
        data = read_page(block_address, page_address);
        if data == None:
            data = [0] * BYTES_PER_PAGE;
        begin_index = byte_address_in_page;
        end_index = min(length + begin_index - 1, BYTES_PER_PAGE - 1);
        read_data.extend(data[begin_index:end_index + 1]);
        address += (BYTES_PER_PAGE - begin_index);
        length -= (BYTES_PER_PAGE - begin_index);
    return read_data;

# reads a single page from memory
def read_page(block_address, page_address):
    check_address('block', block_address);
    check_address('page', page_address);
    page_location = page_map_get(block_address, page_address);
    if page_location == None:
        return None;
    log.append('page read');
    block_address, page_address = page_location;
    return blocks[block_address][page_address][:];

# performs a write to memory using an arbitrary array of data starting from any byte address
def write(address, data, avoid=None):
    check_address('memory', address, len(data));
    while len(data) > 0:
        block_address = address // BYTES_PER_BLOCK;
        byte_address_in_block = address % BYTES_PER_BLOCK;
        page_address = byte_address_in_block // BYTES_PER_PAGE;
        byte_address_in_page = byte_address_in_block % BYTES_PER_PAGE;
        begin_index = byte_address_in_page;
        end_index = min(len(data) + begin_index - 1, BYTES_PER_PAGE - 1);
        if begin_index != 0 or end_index != BYTES_PER_PAGE:
            page = read_page(block_address, page_address);
            if page != None:
                page[begin_index:end_index + 1] = data[0:end_index - begin_index + 1];
                physical_address = page_map_get(block_address, page_address);
                page_state_store(physical_address[0], physical_address[1], PageState.INVALID);
                check_reset_block(physical_address[0]);
            else:
                page = [0] * BYTES_PER_PAGE;
                page[begin_index:end_index + 1] = data[0:end_index - begin_index + 1];
        else:
            page = data[0:BYTES_PER_PAGE + 1];
        store_location = find_available_page(avoid=avoid);
        page_map_store(block_address, page_address, store_location);
        write_page(block_address, page_address, page);
        address += (BYTES_PER_PAGE - begin_index);
        data = data[end_index - begin_index + 1:];
    return;

# writes a single page to memory
def write_page(block_address, page_address, page):
    check_address('block', block_address);
    check_address('page', page_address);
    block_address, page_address = page_map_get(block_address,page_address);
    log.append('page write');
    blocks[block_address][page_address] = page;
    page_state_store(block_address, page_address, PageState.IN_USE);
    return;

# finds an available location in physical memory for a page to be stored
def find_available_page(avoid=None, collect_garbage=True):
    log.append('page search');
    for block_num, block in enumerate(physical_page_state):
        if block_num == avoid:
            continue;
        for page_num, state in enumerate(block):
            if state == PageState.AVAILABLE:
                return (block_num, page_num);
    if collect_garbage:
        garbage_collection();
        find_available_page(collect_garbage=False);
    raise RuntimeError('no available pages');

# moves things around to create an eraseable block
def garbage_collection():
    # find block with the most invalid pages
    max_invalid_pages = 0;
    garbage_block_address = 0;
    for block_address in range(NUM_BLOCKS):
        invalid_pages = count_invalid_pages(block_address);
        max_invalid_pages = max(max_invalid_pages, invalid_pages);
        if max_invalid_pages == invalid_pages:
            garbage_block_address = block_address;
    # use block number garbage_block_address and write its valid pages to other locations
    for page_address in range(PAGES_PER_BLOCK):
        state = page_state_get(garbage_block_address, page_address);
        if state == PageState.IN_USE:
            data = blocks[garbage_block_address][page_address];
            old_address = get_virtual_page((garbage_block_address, page_address));
            write(old_address, data, avoid=garbage_block_address);
        if state != PageState.INVALID:
            page_state_store(garbage_block_address, page_address, PageState.INVALID);
    # clear the garbage block
    check_reset_block(garbage_block_address);
    return;

# returns the byte address of the virtual page that is currently mapped to a given physical page
def get_virtual_page(physical_location):
    for block_address in range(NUM_BLOCKS):
        for page_address in range(PAGES_PER_BLOCK):
            if page_map[block_address][page_address] == physical_location:
                return BYTES_PER_BLOCK*block_address + BYTES_PER_PAGE*page_address;
    raise RuntimeError('Mapping does not exist');

# returns the number of invalid pages in a physical block
def count_invalid_pages(block_address):
    invalid_pages = 0;
    for page_address, page in enumerate(blocks[block_address]):
        if page_state_get(block_address, page_address) == PageState.INVALID:
            invalid_pages += 1;
    return invalid_pages;

# resets a block if it needs to be reset (if the whole block is INVALID)
def check_reset_block(block_address):
    for i in range(PAGES_PER_BLOCK):
        if page_state_get(block_address, i) != PageState.INVALID:
            return;
    erase(block_address);
    return;

# performs an erase of a single block
def erase(block_address):
    check_address('block', block_address);
    log.append('erase');
    for i in range(PAGES_PER_BLOCK):
        blocks[block_address][i] = [0] * BYTES_PER_PAGE;
        page_state_store(block_address, i, PageState.AVAILABLE);
    return;

# checks an address to see if it is within the range of the container it is addressing
def check_address(address_type, address, length=1):
    if address_type == 'block':
        assert 0 <= address and address < NUM_BLOCKS;
    elif address_type == 'page':
        assert 0 <= address and address < PAGES_PER_BLOCK;
    elif address_type == 'memory':
        assert 0 <= address and address + length <= MEMORY_CAPACITY;
    else:
        assert False;
    return;

# gets a location out of the page map
def page_map_get(block_address, page_address):
    #log.append('check page map');
    return page_map[block_address][page_address];

# stores a location in the page map
def page_map_store(block_address, page_address, value):
    #log.append('update page map');
    page_map[block_address][page_address] = value;
    return;

# gets the state of a physical page
def page_state_get(block_address, page_address):
    #log.append('check page state');
    return physical_page_state[block_address][page_address];

# stores a state for a physical page
def page_state_store(block_address, page_address, state):
    #log.append('update page state');
    physical_page_state[block_address][page_address] = state;
    return;

# prints the state of the SSD
def print_SSD(what='all'):
    if what in ['all', 'mem','memory']:
        print_memory();
        print
    if what in ['all', 'state']:
        print_page_state();
        print
    if what in ['all', 'map']:
        print_page_map();
    return;

# prints the page map
def print_page_map():
    print 'PAGE MAP'
    for block_num, block in enumerate(page_map):
        print 'BLOCK ' + str(block_num)
        for page_num, mapping in enumerate(block):
            print '  PAGE ' + str(page_num) + ': ' + str(mapping)

# prints the page state
def print_page_state():
    print 'PHYSICAL PAGE STATE'
    for block_num, block in enumerate(physical_page_state):
        print 'BLOCK ' + str(block_num)
        for page_num, state in enumerate(block):
            print '  PAGE ' + str(page_num) + ': ' + state

# prints the physical memory
def print_memory():
    print 'PHYSICAL MEMORY'
    for block_num, block in enumerate(blocks):
        print 'BLOCK ' + str(block_num)
        for page_num, page in enumerate(block):
            sys.stdout.write('  PAGE ' + str(page_num));
            for x in page:
                sys.stdout.write('\t' + str(x));
            print

# initializes the memory with certain parameters
def init(num_blocks, pages_per_block, bytes_per_page):
    global NUM_BLOCKS, PAGES_PER_BLOCK, BYTES_PER_PAGE, BYTES_PER_BLOCK, MEMORY_CAPACITY, blocks, physical_page_state, page_map;
    NUM_BLOCKS = num_blocks;
    PAGES_PER_BLOCK = pages_per_block;
    BYTES_PER_PAGE = bytes_per_page;
    BYTES_PER_BLOCK = BYTES_PER_PAGE * PAGES_PER_BLOCK;
    MEMORY_CAPACITY = BYTES_PER_BLOCK * NUM_BLOCKS;
    blocks = [[[0] * BYTES_PER_PAGE for _ in range(PAGES_PER_BLOCK)] for __ in range(NUM_BLOCKS)];
    physical_page_state = [[PageState.AVAILABLE] * PAGES_PER_BLOCK for _ in range(NUM_BLOCKS)];
    page_map = [[None] * PAGES_PER_BLOCK for _ in range(NUM_BLOCKS)]; 

# runs the memory through a series of tests specified by a filename
def test(filename):
    with open(filename, 'r') as testfile:
        num_blocks = int(re.sub(r'.*=', '', testfile.readline()).strip());
        pages_per_block = int(re.sub(r'.*=', '', testfile.readline()).strip());
        bytes_per_page = int(re.sub(r'.*=', '', testfile.readline()).strip());
        init(num_blocks, pages_per_block, bytes_per_page);
        for command in testfile.readlines():
            execute(command);

# runs an i/o operation on the SSD
def main(operation, arg1, arg2):
    if operation == 'read':
        print read(int(arg1), int(arg2));
    elif operation == 'write':
        write(int(arg1), arg2);
    else:
        raise RuntimeError
    print log

# executes an SSD operation from a user command
def execute(command):
    del log[:]
    if command.startswith('read'):
        args = command.split(' ');
        print read(int(args[1]), int(args[2]));
        print log
    elif command.startswith('write'):
        args = command.split(' ');
        if args[2].startswith('random'):
            length = int(args[2].replace('random(','').replace(')',''));
            data = random.sample(range(256), length);
        else:
            data = ast.literal_eval(args[2]);
        write(int(args[1]), data);
        print log
    elif command.startswith('print') or command == 'p':
        args = (command + ' all').split(' ');
        print_SSD(args[1]);
    elif command in ['exit', 'e', 'q', 'quit']:
        exit();
    elif command == 'garbage':
        garbage_collection();
    elif command.startswith('expect'):
        args = command.split(' ');
        if read(int(args[1]), int(args[2])) != ast.literal_eval(args[3]):
            print 'FAILED'
            exit();
        else:
            print 'Success: ' + command
    else:
        print 'Invalid command'
    return;

usage="""
Usage: 
 To run a series of tests, use
   python ssd.py test <testfile name>

 To run a single i/o operation, use
   python ssd.py [read|write] <i/o parameters>

 To run a single i/o operation with specified parameters, use
   python ssd.py [read|write] <i/o parameters> <num_blocks> <pages_per_block> <bytes_per_page>

"""
if __name__ == '__main__':
    if len(sys.argv) == 3 and sys.argv[1] == 'test':
        test(sys.argv[2]);
    elif len(sys.argv) == 4:
        init(NUM_BLOCKS, PAGES_PER_BLOCK, BYTES_PER_PAGE);
        main(sys.argv[1], sys.argv[2], sys.argv[3]);
    elif len(sys.argv) == 7:
        init(int(sys.argv[4]), int(sys.argv[5]), int(sys.argv[6]));
        main(sys.argv[1], sys.argv[2], sys.argv[3]);
    else:
        print usage;
        raise RuntimeError

File: ssd.cc
#include <iostream>
#include <stdint.h>
#include <stdio.h>
#include <cstdlib>
#include <string>

// parameters copied from blockdev.h
// TODO: include properly
#define SECTOR_SIZE 32 //512 // bytes per sector
const int BYTES_PER_SECTOR = SECTOR_SIZE;
#define SECTOR_SHIFT 9
#define SECTOR_BEATS (SECTOR_SIZE / 8) // = 64 64-bit (8-byte) chunks per sector
const int INTS_PER_SECTOR = SECTOR_BEATS;
#define MAX_REQ_LEN 16 // sectors per request
FILE *_file;
const char * filename = "disk.dat";
std::string example_data = "The quick brown fox jumps over the lazy dog. If Tid the thistle sifter lifts his sieve of sifted thistles, he'll shift the thistles in his sifting sieve to live unsifted.";

// ssd parameters and convenience variables
const int NUM_BLOCKS = 4; // total number of blocks in the SSD
const int PAGES_PER_BLOCK = 4; // number of pages in each block
const int INTS_PER_PAGE = 8; // number of 64-bit integers in each page
const int BYTES_PER_INT = 8; // number of bytes in a 64-bit integer
const int BYTES_PER_PAGE = INTS_PER_PAGE * BYTES_PER_INT; // total bytes in 1 page
const int BYTES_PER_BLOCK = BYTES_PER_PAGE * PAGES_PER_BLOCK; // total bytes in 1 block
const int MEMORY_CAPACITY_INTS = INTS_PER_PAGE * PAGES_PER_BLOCK; // total storage in ints
const int MEMORY_CAPACITY_BYTES = BYTES_PER_BLOCK * NUM_BLOCKS; // total storage in bytes
const int SECTORS_PER_PAGE = INTS_PER_PAGE / INTS_PER_SECTOR; // number of sectors in each page

/*
 * An enum to represent the state of a page on the disk.
 */
enum class PageState { AVAILABLE=0, IN_USE=1, INVALID=2 };
const char * name[] = {"AVAILABLE", "IN_USE", "INVALID"};

/*
 * Holds an address for a page in the SSD. This is a block address and a page
 * address within that block. It can be a physical address or a virtual address.
 */
struct ssd_address
{
    uint32_t block_address;
    uint32_t page_address;

    ssd_address()
    {
        block_address = -1;
        page_address = -1;
    }
};

/*
 * Maps virtual pages to physical pages. For example, specifying page_map[2][3]
 * will return the physical page that is backing virtual page 3 of virtual
 * block 2.
 */
ssd_address page_map[NUM_BLOCKS][PAGES_PER_BLOCK];

/*
 * Stores the state of each virtual page. If a page is invalid, it means that
 * the data in it is old. The page map should not direct any virtual pages to an
 * invalid physical page. If a page is in use, that means the page map is
 * currently mapping a virtual page to it. If a page is available, that means it
 * has been cleared and is waiting for the page map to start using it.
 */
PageState physical_page_state[NUM_BLOCKS][PAGES_PER_BLOCK];

/*
 * Prints a string with a newline.
 */
template<class T>
void p(T string)
{
    std::cout << string << "\n";
}

/*
 * Defines the min function
 */
int min(int a, int b)
{
    if(a <= b)
        return a;
    return b;
}

/*
 * Prints the first N elements of an array.
 */
void print_array(uint64_t data[], int n, bool characters=false)
{
    printf("[");
    for(int i=0; i<n-1; i++)
    {
        printf(characters?"%c, ":"%d, ", (int) data[i]); 
    }
    printf(characters?"%c]\n":"%d]\n", (int) data[n]);
}

/*
 * Takes a physical ssd_address and returns its offset in bytes.
 */
uint32_t byte_offset_of(ssd_address physical_address)
{
    return physical_address.block_address * BYTES_PER_BLOCK + physical_address.page_address * BYTES_PER_PAGE;
}

/*
 * Takes a byte offset and returns its corresponding ssd address
 */
ssd_address address_of(uint64_t offset)
{
    ssd_address address;
    address.block_address = offset / BYTES_PER_BLOCK;
    address.page_address = (offset % BYTES_PER_BLOCK) / BYTES_PER_PAGE;
    return address;
}

/*
 * Reads data from the file and fills it into a buffer.
 * offset: number of bytes from the beginning of the file.
 * buffer: array that the resulting data gets put into.
 * length: number of sectors to read.
 */
void from_disk(uint64_t offset, uint64_t buffer[], uint32_t length)
{
    int error_code = fseek(_file, offset, SEEK_SET);
    if(error_code)
        p("Could not seek.");
    int sectors_read = fread(buffer, BYTES_PER_SECTOR, length, _file);
    if(sectors_read < length)
        p("Didn't read enough sectors.");
}

/*
 * Writes data from a buffer to the file.
 * offset: number of bytes from the beginning of the file.
 * data: array that the data comes from.
 * count: number of 64-bit integers being written.
 */
void to_disk(uint64_t offset, uint64_t data[], uint64_t count)
{
    int error_code = fseek(_file, offset, SEEK_SET);
    if(error_code)
        p("Could not seek.");
    int sectors_written = fwrite(data, sizeof(uint64_t), count, _file);
    if(sectors_written < count)
        p("Didn't write enough sectors.");
}

/*
 * Gets the state of a physical page out of the state status array.
 */
PageState page_state_get(ssd_address physical_address)
{
    return physical_page_state[physical_address.block_address][physical_address.page_address];
}

/*
 * Stores the state of a physical page in the physical page state status array.
 */
void page_state_store(ssd_address physical_address, PageState state)
{
    physical_page_state[physical_address.block_address][physical_address.page_address] = state;
}

/*
 * Gets the physical page backing a virtual page.
 */
ssd_address page_map_get(ssd_address virtual_address)
{
    return page_map[virtual_address.block_address][virtual_address.page_address];
}

/*
 * Stores the physical location of a virtual page in the page map.
 */
void page_map_store(ssd_address virtual_address, ssd_address physical_address)
{
    page_map[virtual_address.block_address][virtual_address.page_address] = physical_address;
}

/*
 * Looks through the SSD to find a page that is marked as AVAILABLE.
 */
ssd_address find_available_page(uint32_t avoid)
{
    ssd_address store_location;
    store_location.block_address = 0;
    store_location.page_address = 0;
    for(int block_num=0; block_num<NUM_BLOCKS; block_num++)
    {
        if(block_num == avoid)
            continue;
        for(int page_num=0; page_num<PAGES_PER_BLOCK; page_num++)
        {
            store_location.block_address = block_num;
            store_location.page_address = page_num;
            if(page_state_get(store_location) == PageState::AVAILABLE)
                return store_location;
        }
    }
}

/*
 * Writes a single page to memory. The length of the data array should be the
 * length of 1 page.
 */
void write_page(ssd_address virtual_address, uint64_t data[], uint32_t avoid=-1)
{
    ssd_address physical_address = find_available_page(avoid);
    page_map_store(virtual_address, physical_address);
    to_disk(byte_offset_of(physical_address), data, INTS_PER_PAGE);
    p("page write");
    page_state_store(physical_address, PageState::IN_USE);
}

/*
 * Reads a single page from memory.
 */
void read_page(uint64_t buffer[], ssd_address virtual_address)
{
    ssd_address physical_address = page_map_get(virtual_address);
    from_disk(byte_offset_of(physical_address), buffer, SECTORS_PER_PAGE);
    p("page read");
}

/*
 * Erases a block.
 */
void erase_block(uint64_t block_address)
{
    p("erase");
    ssd_address address;
    address.block_address = block_address;
    for(int i=0; i<PAGES_PER_BLOCK; i++)
    {
        address.page_address = i;
        page_state_store(address, PageState::AVAILABLE);
    }
}

/*
 * Checks if the given block needs to be reset and resets it if necessary.
 */
void check_reset_block(uint64_t block_address)
{
    ssd_address address;
    address.block_address = block_address;
    for(int i=0; i<PAGES_PER_BLOCK; i++)
    {
        address.page_address = i;
        if(page_state_get(address) != PageState::INVALID)
            return;
    }
    erase_block(block_address);
}

/*
 * Reads from the memory the way the block device should, ignoring the way the
 * SSD works. Interfaces with the block device model.
 * offset: number of bytes into the memory
 * buffer: location for the data to end up. Must be at least COUNT ints long, rounded up to the nearest whole page
 * count: number of ints to be read
 */
void read(uint64_t offset, uint64_t buffer[], int count)
{
    while(count > 0)
    {
        ssd_address virtual_address = address_of(offset);
        read_page(buffer, virtual_address);
        count -= INTS_PER_PAGE;
        buffer += INTS_PER_PAGE;
        offset += BYTES_PER_PAGE;
    }
}

/*
 * Writes to the memory the way the block device should, ignoring the way the
 * SSD works. Interfaces with the block device model.
 * offset: number of bytes into the memory
 * data: array of ints to be written
 * count: number of ints to be written
 */
void write(uint64_t offset, uint64_t data[], int count)
{
    //while there's still data to write
    while(count > 0)
    {
        ssd_address virtual_address = address_of(offset);
        ssd_address physical_address = page_map_get(virtual_address);
        uint64_t page_offset = offset % BYTES_PER_PAGE;
        uint64_t page[INTS_PER_PAGE];
        p("count");
        p(count);
        p("addresses (v, p)");
        p(virtual_address.block_address);
        p(virtual_address.page_address);
        p(physical_address.block_address);
        p(physical_address.page_address);
        //if there was data there already, read it and merge it with the new data
        if(physical_address.block_address != -1 && page_state_get(physical_address) == PageState::IN_USE)
        {
            p("page was in use");
            read_page(page, virtual_address);
            //mark the old physical location as invalid
            page_state_store(physical_address, PageState::INVALID);
            check_reset_block(physical_address.block_address);
        }
        for(int i=page_offset; i<min(page_offset + count, INTS_PER_PAGE); i++)
        {
            page[i] = data[i];
        }
        write_page(virtual_address, page);
        count -= INTS_PER_PAGE;
        data += INTS_PER_PAGE;
        offset += BYTES_PER_PAGE;
    }
}

/*
 * Prints everything.
 */
void print_ssd()
{
    p("PHYSICAL MEMORY");
    //print the contents of the memory
    for (int block_num=0; block_num<NUM_BLOCKS; block_num++)
    {
        printf("BLOCK %d   SECTORS\n", block_num);
        for (int page_num=0; page_num<PAGES_PER_BLOCK; page_num++)
        {
            printf("  PAGE %d  ", page_num);
            for (int sector_num=0; sector_num<SECTORS_PER_PAGE; sector_num++)
            {
                uint64_t sector[INTS_PER_SECTOR];
                from_disk(block_num*BYTES_PER_BLOCK+page_num*BYTES_PER_PAGE+sector_num*BYTES_PER_SECTOR, sector, 1);
                for (int i=0; i<INTS_PER_SECTOR; i++)
                    printf("%c", (char)sector[i]);
                printf("\t");
            }
            p("");
        }
    }
    p("");
    p("PAGE MAP");
    //print the contents of the page map
    for (int block_num=0; block_num<NUM_BLOCKS; block_num++)
    {
        printf("BLOCK %d\n", block_num);
        for (int page_num=0; page_num<PAGES_PER_BLOCK; page_num++)
        {
            ssd_address loc = page_map[block_num][page_num];
            printf("  PAGE %d: (%d, %d)\n", page_num, loc.block_address, loc.page_address);
        }
    }
    p("");
    p("PAGE STATE");
    //print the contents of the page state status array
    for (int block_num=0; block_num<NUM_BLOCKS; block_num++)
    {
        printf("BLOCK %d\n", block_num);
        for (int page_num=0; page_num<PAGES_PER_BLOCK; page_num++)
        {
            PageState state = physical_page_state[block_num][page_num];
            printf("  PAGE %d: %s\n", page_num, name[static_cast<int>(state)]);
        }
    }
}

//writes a bunch of %-signs to the memory to "clear" it
void clear_mem()
{
    int write_length = NUM_BLOCKS*PAGES_PER_BLOCK*INTS_PER_PAGE;
    uint64_t data[sizeof(uint64_t) * write_length];
    for(int i=0; i<write_length; i++)
        data[i] = 37;
    to_disk(0, data, write_length);
}

//basic test of to_disk and from_disk.
void testing_1()
{
    int write_length = example_data.length();
    printf("Writing example data, length = %d\n", write_length);
    uint64_t write_data[sizeof(uint64_t) * write_length];
    for (int i=0; i<write_length; i++)
        write_data[i] = example_data[i];
    to_disk(0, write_data, write_length);

    uint64_t read_data[MAX_REQ_LEN * INTS_PER_SECTOR];
    int sectors_to_read = 2;
    printf("Reading %d sector(s): ", sectors_to_read);
    from_disk(0, read_data, sectors_to_read);
    for (int i=0; i<sectors_to_read*INTS_PER_SECTOR; i++)
        printf("%c", (int)read_data[i]);
    p("");
}

//basic test of read_page and write_page
void testing_2()
{
    ssd_address address;
    address.block_address = 0;
    address.page_address = 1;
    printf("Created address (%d,%d)\n", address.block_address, address.page_address);
    uint64_t write_data[INTS_PER_PAGE];
    for (int i=0; i<INTS_PER_PAGE; i++)
        write_data[i] = example_data[i];
    printf("Filled write buffer: ");
    for(int i=0; i<INTS_PER_PAGE; i++)
        printf("%c", (int)write_data[i]);
    p("");
    write_page(address, write_data);

    address.page_address = 1;
    int read_length = INTS_PER_PAGE;
    uint64_t read_data[sizeof(uint64_t) * read_length];
    read_page(read_data, address);
    printf("Read data from (%d,%d): ", address.block_address, address.page_address);
    for(int i=0; i<read_length; i++)
        printf("%c", (int)read_data[i]);
    p("");
}

//basic test of read() and write()
void testing_3()
{
    //write new data to the same location
    uint64_t write_data[5];
    int length = 5;
    write_data[0] = 67;
    write_data[1] = 67;
    write_data[2] = 67;
    write_data[3] = 67;
    write_data[4] = 67;
    write(0, write_data, length);
    write_data[0] = 68;
    write_data[1] = 68;
    write_data[2] = 68;
    write_data[3] = 68;
    write_data[4] = 68;
    write(0, write_data, length);
    write(BYTES_PER_PAGE * 2, write_data, length);

    //write many pages at once
    uint64_t write_data_long[40];
    length = 40;
    for(int i=0; i<length; i++)
    {
        write_data_long[i] = i + 55;
    }
    write(BYTES_PER_PAGE * 5, write_data_long, length);

    //read
    uint64_t read_data[8];
    read(0, read_data, 8);
    print_array(read_data, 8, true);

    //read many pages at once
    uint64_t read_data_long[56];
    read(0, read_data_long, 56);
    print_array(read_data_long, 56, true);
    //TODO: what happens if you want to read invalid/unwritten data? I guess it all needs to be zeroed out.
}

int main(int argc, char* argv[])
{
    _file = fopen(filename, "r+");

    //testing_1();
    //clear_mem();
    //testing_2();
    testing_3();
    print_ssd();

    return 0;
}
